### PR TITLE
WIP: Allow to select if mute indicator is based on standard sink or source (microphone)

### PR DESCRIPTION
**Pull request by penguineer**
I wanted to change the mute indicator to show my microphone state (alongside the change in my window manager to mute the microphone when I press the mute button). With this change the source can be configured in the keyboard profile.

Unfortunately I cannot test this on a MAC, so the feature is just unavailable there.

I leave it open for discussion if the source setting should be based on the profile or in the general settings.

### CODE REVIEW

**Comment by penguineer** [comment]
I wonder if parallel to the indicator source the mute button could also be rewired on the ckb-next side to either mute the microphone or the speakers. Currently I do this in my window manager.

**Comment by ZirconiumX** [comment]
Since it's no longer WIP would you like us to review it?

File: src/gui/kbperf.cpp [hunk 1]
```diff
@@ -132,6 +133,10 @@ void KbPerf::load(CkbSettings& settings){
         }
     }
     SGroup group(settings, "Performance");
+    {
+        QString s_muteDev = settings.value("MuteDevice", "sink").toString();
+        iMuteDev = s_muteDev == "source" ? SOURCE : SINK;
```
Comment (tatokis) [inline src/gui/kbperf.cpp:138]: Please use parentheses around ternary operators.
Makes them easier to read.

File: src/gui/kbperf.cpp [hunk 2]
```diff
@@ -211,6 +216,7 @@ void KbPerf::load(CkbSettings& settings){
 void KbPerf::save(CkbSettings& settings){
     _needsSave = false;
     SGroup group(settings, "Performance");
+    settings.setValue("MuteDevice", iMuteDev == SOURCE ? "source" : "sink");
```
Comment (tatokis) [inline src/gui/kbperf.cpp:219]: Same as above

File: src/gui/kperfwidget.cpp [hunk 3]
```diff
@@ -57,6 +57,16 @@ KPerfWidget::KPerfWidget(QWidget *parent) :
             });
         }
     }
+    muteDev = ui->muteDev;
+//    muteDev->setCurrentIndex(perf ? perf->getMuteDevice() : 1);
```
Comment (ZirconiumX) [inline src/gui/kperfwidget.cpp:61]: Can you clean this comment up please?

File: src/gui/kperfwidget.cpp [hunk 4]
```diff
@@ -57,6 +57,16 @@ KPerfWidget::KPerfWidget(QWidget *parent) :
...
+    if (isMuteDeviceSupported())  {
+        muteDev->setEnabled(true);
+        connect(muteDev, QOverload<int>::of(&QComboBox::currentIndexChanged), [=] (int index) {
```
Comment (tatokis) [inline src/gui/kperfwidget.cpp:64]: We have a Q_OVERLOAD macro, please use that as it retains compatibility with older Qt versions that we still support.

File: src/gui/kperfwidget.ui [hunk 5]
```diff
@@ -94,6 +94,20 @@
      </property>
     </widget>
    </item>
+   <item row="18" column="4" colspan="8">
+    <widget class="QComboBox" name="muteDev">
+     <item>
+      <property name="text">
+       <string>Use default sink (Audio)</string>
```
Comment (tatokis) [inline src/gui/kperfwidget.ui:101]: These are not user-friendly names.
I'd suggest "Use default output device"

File: src/gui/kperfwidget.ui [hunk 6]
```diff
@@ -94,6 +94,20 @@
...
+      </property>
+     </item>
+     <item>
+       <property name="text">
+        <string>Use default source (Microphone)</string>
```
Comment (tatokis) [inline src/gui/kperfwidget.ui:106]: Possibly "Use default input device"

File: src/gui/media_linux.cpp [hunk 7]
```diff
@@ -33,16 +35,26 @@ static void SinkCallback(pa_context* context, const pa_sink_info* info, int eol,
...
+    lastKnownSink = info->mute ? MUTED : UNMUTED;
+}
+
+static void SourceCallback(pa_context* context, const pa_source_info* info, int eol, void* data){
+    QMutexLocker locker(&mutex);
+
+    if(info == nullptr || info->name != defaultSource)
+        return;
+
+    lastKnownSource = info->mute ? MUTED : UNMUTED;
```
Comment (tatokis) [inline src/gui/media_linux.cpp:47]: Parentheses around here would be appreciated

File: src/gui/media_linux.cpp [hunk 8]
```diff
@@ -102,7 +124,15 @@ static void ContextStateCallback(pa_context* context, void* data){
 
 #endif
 
-muteState getMuteState(){
+bool isMuteDeviceSupported() {
+    return true;
+}
+
+muteState mute_state(const muteDevice muteDev) {
+    return muteDev == SINK ? lastKnownSink : lastKnownSource;
```
Comment (tatokis) [inline src/gui/media_linux.cpp:132]: Parentheses

File: src/gui/kperfwidget.cpp (hunk 4 above)
Comment (tatokis) [reply src/gui/kperfwidget.cpp:64]: Sorry, it's called `OVERLOAD_PTR` https://github.com/ckb-next/ckb-next/blob/050231f2560fdfa25668cea1a57301dc79b944d4/src/gui/qoverloadlegacy.h

**Comment by penguineer** [comment]
Thank you for the feedback! I will check and push the changes.

File: src/gui/kbperf.cpp (hunk 1 above)
Comment (penguineer) [reply src/gui/kbperf.cpp:138]: I understand that you would like to see `(s_muteDev == "source") ? SOURCE : SINK`

**Comment by penguineer** [comment]
I have fixed the issues. Please let me know if you would like them  squashed into the original commits.

**Comment by penguineer** [comment]
It seems that the setting is not copied when the profile is duplicated. I need to fix this.

**Comment by penguineer** [comment]
What can I do to help with this PR?

**Comment by tatokis** [comment]
Sorry, I've been busy with other things. 
I didn't like how the ui was laid out and I was faking to suggest some changes that required changing other parts of the gui well. 

I made a mockup which I'll share later tonight once I find it.

**Comment by penguineer** [comment]
> Sorry, I've been busy with other things. I didn't like how the ui was laid out and I was faking to suggest some changes that required changing other parts of the gui well. I made a mockup which I'll share later tonight once I find it.

Okay, thank you for the quick reply!

I am open  to suggestions for the UI.

**Comment by tatokis** [comment]
I was thinking something like this
7cd6aa4a2edb

**Comment by penguineer** [comment]
If I understand your mock-up correctly, there is a decision to be made:

A) Use your proposal, which is visually more appealing, because it brings more order to the way the dialog is laid out.
B) Use my proposal which adheres to the UX principle to keep similar functions (configuration for the mute button) close together.

Which way do we go?

**Comment by tatokis** [comment]
A month later... (Sorry.)
After thinking about it, you are right. Unfortunately both aren't great solutions.
I do prefer my suggestion, as it doesn't break layout, as I tend to get annoyed when I see bad layouts, but it isn't intuitive.

I do think that consistent layout was the thought process when the UI was originally designed by ccMSC, as the "indicator intensity" option wasn't placed under "Top row" but rather under Miscellaneous, on its own.

**Comment by penguineer** [comment]
Okay, I will change it to your suggestion then.

In case we have more discussion about the UI layout, I would liker to put the "backend" part, which we agree on, in a distinct PR to get it merged. Would that be an option for you?

**Comment by tatokis** [comment]
I'm not sure I understand correctly. Wouldn't doing that add dead code?

**Comment by penguineer** [comment]
> I'm not sure I understand correctly. Wouldn't doing that add dead code?

More like "dormant" code. It still extends the code base and  gives others the chance to use or extend it. But it means that the back-end part can already be taken into account during changes while we figure out the front-end.

**Comment by tatokis** [comment]
Sure, go ahead

**Comment by penguineer** [comment]
> Sure, go ahead

Thank you, I've submitted #661

**Comment by penguineer** [comment]
I created a different layout based on your suggestion. Please have a look. :)

**Comment by tatokis** [comment]
So, apologies for this (as it was my suggestion). I just had a look and I feel that the "copy to mode" button is a bit confusing there.

Perhaps it would be better if it was moved below "Indicate mute from", and renamed to something like "Copy performance settings to mode..." or something along those lines.

**Comment by tatokis** [comment]
https://github.com/ckb-next/ckb-next/commit/ca7c2ba010012f1d8c98cf084efb405369e05aa9

**Pull request closed without merging**